File: Rust/linkedList/singleLinkedList.rs
#![allow(non_snake_case)]

#[derive(Debug)]
pub struct SingleLinkedList<T: 'static> {
    length: u32, 
    head: Next<T>,
}

#[derive(Debug)]
struct Node<T: 'static> {
    data: T,
    next: &'static mut Next<T>,
}

#[derive(Debug)]
enum Next<T: 'static> {
    Some(Node<T>),
    None
}

impl<T: std::fmt::Debug> SingleLinkedList<T> {
    pub fn new() -> SingleLinkedList<T> {
        SingleLinkedList { length: 0, head: Next::None }
    }

    pub fn length(&self) -> u32 {
        return self.length;
    }

    pub fn appendNode(&mut self, data: T) {

        let newNode = Node::<T> { data: data, next: &mut Next::None };

        // self.head = Next::Some(newNode);
        self.length += 1;

        let mut curNode = &mut self.head;

        match curNode {
            Next::None => self.head = Next::Some(newNode),
            Next::Some(node) => curNode = node.next,
        }

        // loop {
        //     match curNode {
        //         Next::None => {
        //             self.head = Next::Some(newNode);
        //             break;
        //         },
        //         Next::Some(node) => {

        //         },
        //     }
        // }

        self.length += 1;
        // match curNode {
        //     Next::None => println!("Nothing lmao"),
        //     Next::Some(node) => println!("{:?}", node),
        // }
    }
}

File: C#/linkedList/SingleLinkedList.cs
using System;

namespace DataStructures.SLL;

public class Node<T> {
    public T data;
    public Node<T>? next;

    public Node(T data) {
        this.data = data;
        this.next = null;
    }

    public bool Compare(T other) {
        if(!this.data.Equals(other)) {
            return false;
        }
        return true;
    }
}

public class SingleLinkedList<T> {
    public Node<T> head;
    public int length;

    public SingleLinkedList(){
        this.length = 0;
        this.head = null;
    }

    public void prependNode(T data){
        Node<T> newNode = new Node<T>(data);

        if (this.length == 0) {
            this.head = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }

        this.length++;
    }

    public void appendNode(T data) {
        Node<T> newNode = new Node<T>(data);

        if (this.length == 0) {
            this.head = newNode;
        } else {
            Node<T> curNode = this.head;

            while (curNode.next != null) {
                curNode = curNode.next;
            }

            curNode.next = newNode;
        }

        this.length++;
    }

    public void deleteNode(T data) {
        if (this.length == 0) return;

        Node<T> curNode = this.head;
        if (curNode.Compare(data)) {
            this.head = this.head.next;
            curNode.next = null;
            this.length--;
        }

        while (curNode.next != null) {
            if (curNode.next.Compare(data)) {
                Node<T> deleteNode = curNode.next;
                curNode.next = deleteNode.next;
                deleteNode = null;
                this.length--;
            } else {
                curNode = curNode.next;
            }
        }
    }

    public void deleteAllNodes(T data) {
        if (this.length == 0) {
            return;
        }
        Node<T> curNode = this.head;
        while (this.head.Compare(data)) {
            this.head = this.head.next;
            curNode = this.head;
            this.length--;
            // Stop if the List is empty
            if (curNode == null) return;
        }

        while (curNode.next != null) {
            if(curNode.next.Compare(data)) {
                Node<T> deleteNode = curNode.next;
                curNode.next = deleteNode.next;
                deleteNode = null;
                this.length--;
            } else {
                curNode = curNode.next;
            }
        }
    }

    public Boolean isEmpty() {
        return this.length == 0 ? true : false;
    }

    public Node<T> findNode(T data) {
        Node<T> curNode = this.head;

        while (!curNode.Compare(data)) {
            if(curNode.next == null) {
                return null;
            } else {
                curNode = curNode.next;
            }
        }

        return curNode;
    }

    public Node<T> getNextNode(Node<T> node) {
        if (node is Node<T>) {
            return node.next;
        }
        return null;
    }

    public void printList() {
        Node<T> curNode = this.head;
        Console.WriteLine($"\nPrinting List of Length: {this.length}\n--START--");
        while(curNode != null) {
            Console.WriteLine(curNode.data);
            curNode = curNode.next;
        }
        Console.WriteLine("--FINISH--");
    }
}
